homeassistant/components/rfxtrx/__init__.py

```diff
@@ -246,52 +263,40 @@ def handle_receive(event):
         data_bits = get_device_data_bits(event.device, devices)
         device_id = get_device_id(event.device, data_bits=data_bits)
 
+        # Register new devices
+        if config[CONF_AUTOMATIC_ADD] and device_id not in devices:
+            _add_device(event, device_id)
+
         # Callback to HA registered components.
-        hass.helpers.dispatcher.dispatcher_send(SIGNAL_EVENT, event, device_id)
+        hass.helpers.dispatcher.async_dispatcher_send(SIGNAL_EVENT, event, device_id)
 
         # Signal event to any other listeners
         fire_event = devices.get(device_id, {}).get(CONF_FIRE_EVENT)
         if fire_event:
             hass.bus.fire(EVENT_RFXTRX_EVENT, event_data)
 
     @callback
-    def device_update(event, device_id):
-        if device_id not in devices:
-            data = entry.data.copy()
-            event_code = binascii.hexlify(event.data).decode("ASCII")
-            data[CONF_DEVICES][event_code] = device_id
-            hass.config_entries.async_update_entry(entry=entry, data=data)
-            devices[device_id] = {}
-
-    if config[CONF_AUTOMATIC_ADD]:
-        hass.helpers.dispatcher.async_dispatcher_connect(SIGNAL_EVENT, device_update)
-
-    device = config[CONF_DEVICE]
-    host = config[CONF_HOST]
-    port = config[CONF_PORT]
-    debug = config[CONF_DEBUG]
-
-    if port is not None:
-        # If port is set then we create a TCP connection
-        rfx_object = rfxtrxmod.Connect(
-            (host, port),
-            None,
-            debug=debug,
-            transport_protocol=rfxtrxmod.PyNetworkTransport,
-        )
-    else:
-        rfx_object = rfxtrxmod.Connect(device, None, debug=debug)
+    def _add_device(event, device_id):
+        """Add a device to config entry."""
+        data = entry.data.copy()
+        event_code = binascii.hexlify(event.data).decode("ASCII")
+        data[CONF_DEVICES][event_code] = device_id
+        hass.config_entries.async_update_entry(entry=entry, data=data)
+        devices[device_id] = {}
 
+    @callback
     def _start_rfxtrx(event):
-        rfx_object.event_callback = handle_receive
-
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, _start_rfxtrx)
+        """Start receiving events."""
+        rfx_object.event_callback = lambda event: hass.async_add_job(
```

Comment:
The rfx instance will call the callback from another thread, right? So we need to use `hass.add_job`.

Reply:
right, indeed.